File: raster/src/main/scala/geotrellis/raster/GridDefinition.scala
package geotrellis.raster

import geotrellis.vector.Extent
import scala.math.{min, max, ceil}

trait GridDefinition {
  def extent: Extent
  def cellwidth: Double
  def cellheight: Double

  def cellSize = CellSize(cellwidth, cellheight)

  def toGridExtent: GridExtent = GridExtent(extent, cellwidth, cellheight)

  /** Creates a RasterExtent out of this GridExtent.
    *
    * @note                      Use with caution: if the number of columns or rows are larger than Int.MaxValue,
    *                            this will throw an exception. Also, if columns * rows > Int.MaxValue, this will
    *                            create a RasterExtent for a raster that could not be backed by any of the Array-backed 
    *                            tile types.
    */
  def toRasterExtent(): RasterExtent = {
    val targetCols = math.round(extent.width / cellwidth).toLong
    val targetRows = math.round(extent.height / cellheight).toLong
    if(targetCols > Int.MaxValue) {
      throw new GeoAttrsError(s"Cannot convert GridExtent into a RasterExtent: number of columns exceeds maximum integer value ($targetCols > ${Int.MaxValue})")
    }
    if(targetRows > Int.MaxValue) {
      throw new GeoAttrsError(s"Cannot convert GridExtent into a RasterExtent: number of rows exceeds maximum integer value ($targetRows > ${Int.MaxValue})")
    }

    RasterExtent(extent, cellwidth, cellheight, targetCols.toInt, targetRows.toInt)
  }

  /**
   * Returns a GridExtent that lines up with this grid' resolution
   * and grid layout.
   * i.e., the resulting GridExtent will not have the given extent,
   * but will have the smallest extent such that the whole of
   * the given extent is covered, that lines up with the grid.
   */
  def createAlignedGridExtent(targetExtent: Extent): GridExtent = {
    val xmin = extent.xmin + (math.floor((targetExtent.xmin - extent.xmin) / cellwidth) * cellwidth)
    val xmax = extent.xmax - (math.floor((extent.xmax - targetExtent.xmax) / cellwidth) * cellwidth)
    val ymin = extent.ymin + (math.floor((targetExtent.ymin - extent.ymin) / cellheight) * cellheight)
    val ymax = extent.ymax - (math.floor((extent.ymax - targetExtent.ymax) / cellheight) * cellheight)

    GridExtent(Extent(xmin, ymin, xmax, ymax), cellwidth, cellheight)
  }

  /**
   * Returns a RasterExtent that lines up with this RasterExtent's resolution,
   * and grid layout.
   * i.e., the resulting RasterExtent will not have the given extent,
   * but will have the smallest extent such that the whole of
   * the given extent is covered, that lines up with the grid.
   */
  def createAlignedRasterExtent(targetExtent: Extent): RasterExtent =
    createAlignedGridExtent(targetExtent).toRasterExtent

  /**
    * Gets the Extent that matches the grid bounds passed in, aligned with this RasterExtent.
    * 
    * @param     gridBounds      The extent to get the grid bounds for
    * @param     clamp          Determines whether or not to clamp the Extent to the
    *                           extent of this RasterExtent; defaults to true. If true, the
    *                           returned extent will be contained by this RasterExtent's extent,
    *                           if false, the Extent returned can be outside of this RasterExtent's extent.
    */
  def extentFor(gridBounds: GridBounds, clamp: Boolean = true): Extent = {
    val xmin = gridBounds.colMin * cellwidth + extent.xmin
    val ymax = extent.ymax - (gridBounds.rowMin * cellheight)
    val xmax = xmin + (gridBounds.width * cellwidth)
    val ymin = ymax - (gridBounds.height * cellheight)

    if(clamp) {
      Extent(
        max(min(xmin, extent.xmax), extent.xmin),
        max(min(ymin, extent.ymax), extent.ymin),
        max(min(xmax, extent.xmax), extent.xmin),
        max(min(ymax, extent.ymax), extent.ymin)
      )
    } else {
      Extent(xmin, ymin, xmax, ymax)
    }
  }
}

File: geotrellis-python/geotrellis/raster/GridDefinition.py
from __future__ import absolute_import
from geotrellis.raster.CellSize import CellSize
#from geotrellis.raster.RasterExtent import RasterExtent, GeoAttrsError
from geotrellis.vector.Extent import Extent
#from geotrellis.raster.GridExtent import GridExtent
from geotrellis.raster.package_scala import INTMAX

class GridDefinition(object):
    @property
    def cellSize(self):
        return CellSize(self.cellwidth, self.cellheight)

    @property
    def toGridExtent(self):
        from geotrellis.raster.GridExtent import GridExtent
        return GridExtent(self.extent, self.cellwidth, self.cellheight)

    def toRasterExtent(self):
        from geotrellis.raster.RasterExtent import RasterExtent, GeoAttrsError
        targetCols = round( float(self.extent.width) / self.cellwidth)
        targetRows = round( float(self.extent.height) / self.cellheight)
        if targetCols > INTMAX:
            raise GeoAttrsError(
                    "Cannot convert GridExtent into a RasterExtent:" +
                    " number of columns exceeds maximum integer value " +
                    "({tc} > {mx})".format(tc = targetCols, mx = INTMAX))
        if targetRows > INTMAX:
            raise GeoAttrsError(
                    "Cannot convert GridExtent into a RasterExtent:" +
                    " number of rows exceeds maximum integer value " +
                    "({tr} > {mx})".format(tc = targetRows, mx = INTMAX))
        return RasterExtent(self.extent, self.cellwidth, self.cellheight, targetCols, targetRows)

    def createAlignedGridExtent(self, targetExtent):
        from geotrellis.raster.GridExtent import GridExtent
        xmin = self.extent.xmin + (math.floor((targetExtent.xmin - self.extent.xmin) / self.cellwidth) * self.cellwidth)
        xmax = self.extent.xmax - (math.floor((self.extent.xmax - targetExtent.xmax) / self.cellwidth) * self.cellwidth)
        ymin = self.extent.ymin + (math.floor((targetExtent.ymin - self.extent.ymin) / self.cellheight) * self.cellheight)
        ymax = self.extent.ymax - (math.floor((self.extent.ymax - targetExtent.ymax) / self.cellheight) * self.cellheight)

        return GridExtent(Extent(xmin, ymin, xmax, ymax), self.cellwidth, self.cellheight)

    def createAlignedRasterExtent(self, targetExtent):
        return self.createAlignedGridExtent(targetExtent).toRasterExtent()

    def extentFor(self, gridBounds, clamp = True):
        xmin = gridBounds.colMin * self.cellwidth + self.extent.xmin
        ymax = self.extent.ymax - (gridBounds.rowMin * self.cellheight)
        xmax = xmin + (gridBounds.width * self.cellwidth)
        ymin = ymax - (gridBounds.height * self.cellheight)

        if clamp:
            return Extent(
                    max(min(xmin, self.extent.xmax), self.extent.xmin),
                    max(min(ymin, self.extent.ymax), self.extent.ymin),
                    max(min(xmax, self.extent.xmax), self.extent.xmin),
                    max(min(xmax, self.extent.ymax), self.extent.ymin))
        else:
            return Extent(xmin, ymin, xmax, ymax)
